In [2]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)
df = pd.read_csv("../data/raw/public_transport_delays.csv")

In [3]:
print("Shape:", df.shape)

Shape: (2000, 24)


In [4]:
df.head()

,trip_id,date,time,transport_type,route_id,origin_station,destination_station,scheduled_departure,scheduled_arrival,actual_departure_delay_min,actual_arrival_delay_min,weather_condition,temperature_C,humidity_percent,wind_speed_kmh,precipitation_mm,event_type,event_attendance_est,traffic_congestion_index,holiday,peak_hour,weekday,season,delayed
0,T00000,2023-01-01,05:00:00,Tram,Route_15,Station_31,Station_6,05:02:00,05:55:00,12,3,Storm,5.1,52,46,13.0,NaN,500,81,0,1,6,Winter,0
1,T00001,2023-01-01,05:15:00,Metro,Route_12,Station_49,Station_32,05:16:00,05:55:00,15,9,Rain,34.0,64,11,11.4,NaN,0,53,0,0,6,Autumn,1
2,T00002,2023-01-01,05:30:00,Bus,Route_16,Station_29,Station_42,05:33:00,06:17:00,0,0,Clear,29.5,35,31,14.1,Sports,0,67,1,0,6,Autumn,0
3,T00003,2023-01-01,05:45:00,Tram,Route_19,Station_26,Station_18,05:49:00,06:08:00,15,10,Clear,27.4,55,41,6.4,NaN,500,84,0,0,6,Winter,1
4,T00004,2023-01-01,06:00:00,Tram,Route_8,Station_18,Station_15,06:00:00,06:35:00,-1,14,Snow,0.1,90,30,18.5,NaN,500,46,0,0,6,Spring,1


In [5]:
df.columns.tolist()

['trip_id',
 'date',
 'time',
 'transport_type',
 'route_id',
 'origin_station',
 'destination_station',
 'scheduled_departure',
 'scheduled_arrival',
 'actual_departure_delay_min',
 'actual_arrival_delay_min',
 'weather_condition',
 'temperature_C',
 'humidity_percent',
 'wind_speed_kmh',
 'precipitation_mm',
 'event_type',
 'event_attendance_est',
 'traffic_congestion_index',
 'holiday',
 'peak_hour',
 'weekday',
 'season',
 'delayed']

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 24 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   trip_id                     2000 non-null   str    
 1   date                        2000 non-null   str    
 2   time                        2000 non-null   str    
 3   transport_type              2000 non-null   str    
 4   route_id                    2000 non-null   str    
 5   origin_station              2000 non-null   str    
 6   destination_station         2000 non-null   str    
 7   scheduled_departure         2000 non-null   str    
 8   scheduled_arrival           2000 non-null   str    
 9   actual_departure_delay_min  2000 non-null   int64  
 10  actual_arrival_delay_min    2000 non-null   int64  
 11  weather_condition           2000 non-null   str    
 12  temperature_C               2000 non-null   float64
 13  humidity_percent            2000 non-null   

In [7]:
df.isnull().sum()

trip_id                          0
date                             0
time                             0
transport_type                   0
route_id                         0
origin_station                   0
destination_station              0
scheduled_departure              0
scheduled_arrival                0
actual_departure_delay_min       0
actual_arrival_delay_min         0
weather_condition                0
temperature_C                    0
humidity_percent                 0
wind_speed_kmh                   0
precipitation_mm                 0
event_type                    1173
event_attendance_est             0
traffic_congestion_index         0
holiday                          0
peak_hour                        0
weekday                          0
season                           0
delayed                          0
dtype: int64

In [8]:
df["event_type"].value_counts(dropna=False)

event_type
NaN         1173
Festival     221
Sports       212
Concert      203
Parade       105
Protest       86
Name: count, dtype: int64

In [9]:
## Verifying Hypothesis

In [10]:
df.assign(event_type_filled=df["event_type"].fillna("No_Event")) \
  .groupby("event_type_filled")["event_attendance_est"] \
  .agg(["count", "mean", "min", "max"])

,count,mean,min,max
event_type_filled,,,,
Concert,203,6364.532020,0,50000
Festival,221,8882.352941,0,50000
No_Event,1173,6282.182438,0,50000
Parade,105,4071.428571,0,50000
Protest,86,5622.093023,0,50000
Sports,212,6158.018868,0,50000


In [11]:
df["event_attendance_est"].describe()

count     2000.000000
mean      6420.250000
std      15198.306129
min          0.000000
25%          0.000000
50%          0.000000
75%       2000.000000
max      50000.000000
Name: event_attendance_est, dtype: float64

In [12]:
df.duplicated().sum()

np.int64(0)

In [13]:
df.describe()

,actual_departure_delay_min,actual_arrival_delay_min,temperature_C,humidity_percent,wind_speed_kmh,precipitation_mm,event_attendance_est,traffic_congestion_index,holiday,peak_hour,weekday,delayed
count,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000
mean,8.688000,13.318000,15.121350,64.714000,29.300500,9.860700,6420.250000,50.244000,0.089500,0.272000,2.976000,0.749500
std,6.268118,9.289727,11.479424,20.334747,17.264015,5.781373,15198.306129,29.225751,0.285535,0.445101,1.990328,0.433409
min,-2.000000,-3.000000,-5.000000,30.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,3.000000,5.000000,5.100000,46.000000,15.000000,4.900000,0.000000,25.000000,0.000000,0.000000,1.000000,0.000000
50%,9.000000,13.000000,15.300000,65.000000,29.000000,9.700000,0.000000,50.000000,0.000000,0.000000,3.000000,1.000000
75%,14.000000,21.000000,24.800000,83.000000,45.000000,14.800000,2000.000000,76.000000,0.000000,1.000000,5.000000,1.000000
max,19.000000,29.000000,35.000000,99.000000,59.000000,20.000000,50000.000000,99.000000,1.000000,1.000000,6.000000,1.000000


In [14]:
df["delayed"].value_counts()

delayed
1    1499
0     501
Name: count, dtype: int64

In [15]:
df["transport_type"].value_counts()

transport_type
Bus      525
Train    493
Tram     491
Metro    491
Name: count, dtype: int64

In [16]:
df["weather_condition"].value_counts()

weather_condition
Clear     343
Snow      343
Storm     337
Fog       331
Cloudy    325
Rain      321
Name: count, dtype: int64

In [17]:
df["season"].value_counts()

season
Summer    524
Spring    502
Winter    494
Autumn    480
Name: count, dtype: int64

In [18]:
df["weekday"].value_counts().sort_index()

weekday
0    288
1    288
2    288
3    288
4    288
5    288
6    272
Name: count, dtype: int64

In [19]:
df["route_id"].nunique()

20

In [20]:
df["origin_station"].nunique()

50

In [21]:
df["destination_station"].nunique()

50

In [22]:
df["actual_arrival_delay_min"].value_counts().sort_index()

actual_arrival_delay_min
-3     55
-2     61
-1     49
 0     55
 1     50
 2     57
 3     55
 4     62
 5     57
 6     58
 7     49
 8     45
 9     65
 10    76
 11    84
 12    64
 13    66
 14    67
 15    62
 16    69
 17    66
 18    66
 19    64
 20    51
 21    53
 22    58
 23    69
 24    75
 25    56
 26    61
 27    63
 28    65
 29    47
Name: count, dtype: int64

In [23]:
df[["actual_departure_delay_min", "actual_arrival_delay_min"]].corr()

,actual_departure_delay_min,actual_arrival_delay_min
actual_departure_delay_min,1.000000,-0.007161
actual_arrival_delay_min,-0.007161,1.000000


In [24]:
print(df["date"].min())
print(df["date"].max())

2023-01-01
2023-01-22


In [25]:
df.nunique().sort_values()

delayed                          2
peak_hour                        2
holiday                          2
transport_type                   4
season                           4
event_attendance_est             5
event_type                       5
weather_condition                6
weekday                          7
route_id                        20
actual_departure_delay_min      22
date                            22
actual_arrival_delay_min        33
destination_station             50
origin_station                  50
wind_speed_kmh                  60
humidity_percent                70
time                            96
traffic_congestion_index       100
precipitation_mm               201
temperature_C                  397
scheduled_departure            474
scheduled_arrival             1068
trip_id                       2000
dtype: int64

# Dataset Audit Summary

## Dataset
- 2000 rows
- 24 columns

## Missing Values
- event_type: 1173 missing (~58.6%)
- Likely represents No_Event

## Duplicate Rows
- 0 duplicates

## Target Variable
- actual_arrival_delay_min

## Secondary Target
- delayed

## Features to Drop
- trip_id

## Potential Leakage
- actual_departure_delay_min

## Key Observations
- Balanced transport types
- Balanced weather categories
- Balanced seasons
- 20 routes
- 50 stations
- Date range: 2023-01-01 to 2023-01-22